torchgeo/datasets/geo.py

```diff
@@ -990,35 +953,38 @@ def __init__(
         .. versionadded:: 0.4
             The *transforms* parameter.
         """
-        super().__init__(transforms)
         self.datasets = [dataset1, dataset2]
         self.collate_fn = collate_fn
+        self.transforms = transforms
 
         for ds in self.datasets:
             if not isinstance(ds, GeoDataset):
                 raise ValueError('IntersectionDataset only supports GeoDatasets')
 
-        self.crs = dataset1.crs
-        self.res = dataset1.res
-
-        # Merge dataset indices into a single index
-        self._merge_dataset_indices()
-
-    def _merge_dataset_indices(self) -> None:
-        """Create a new R-tree out of the individual indices from two datasets."""
-        i = 0
-        ds1, ds2 = self.datasets
-        for hit1 in ds1.index.intersection(ds1.index.bounds, objects=True):
-            for hit2 in ds2.index.intersection(hit1.bounds, objects=True):
-                box1 = BoundingBox(*hit1.bounds)
-                box2 = BoundingBox(*hit2.bounds)
-                box3 = box1 & box2
-                # Skip 0 area overlap (unless 0 area dataset)
-                if box3.area > 0 or box1.area == 0 or box2.area == 0:
-                    self.index.insert(i, tuple(box3))
-                    i += 1
-
-        if i == 0:
+        dataset2.crs = dataset1.crs
+        dataset2.res = dataset1.res
+
+        # Spatial intersection
+        index1 = dataset1.index.reset_index()
+        index2 = dataset2.index.reset_index()
+        self.index = gpd.overlay(
+            index1, index2, how='intersection', keep_geom_type=True
+        )
+
+        # Temporal intersection
```

Comment:
TODO: does this actually work? It doesn't seem to be taking an intersection, especially if both datasets aren't sorted by time.

Reply:
Update: it works, but I don't know why. My best guess is that `gpd.overlay` returns all possible intersections, then the valid check filters out all invalid cases.